### PR TITLE
refactor(ecs): Refactor container definition

### PR DESCRIPTION
**Pull request by mmiermans**
# Goal
Improve code quality of the container definition, by setting the properties on a Javascript object, instead of making replacements in a JSON string.

## Todos
- [x] Verify that the only difference in snapshots is whitespace using a [json diff](https://jsoncompare.com/#!/diff/) tool.
- [x] Deploy example.ts to Pocket-Dev
   - Apply complete! Resources: 37 added, 0 changed, 0 destroyed.

### CODE REVIEW

File: src/base/ApplicationECSContainerDefinition.ts [hunk 1]
```diff
@@ -92,81 +36,66 @@ export interface ApplicationECSContainerDefinitionProps {
...
   if (config.command) {
-    templateInstance = templateInstance.replace(
-      '[[COMMAND]]',
-      '[' + config.command.map((c) => `"${c}"`).join(',') + ']'
-    );
-  } else {
-    templateInstance = templateInstance.replace('"command": [[COMMAND]],', '');
-  }
-
-  templateInstance = templateInstance.replace('[[LOG_GROUP]]', config.logGroup);
```
Comment (mmiermans) [inline src/base/ApplicationECSContainerDefinition.ts:112]: Question: `logGroup` is optional, but this code will fail if it's not set, right? Should it be made required?

File: src/base/ApplicationECSContainerDefinition.ts [hunk 2]
```diff
@@ -92,81 +35,66 @@ export interface ApplicationECSContainerDefinitionProps {
...
   config: ApplicationECSContainerDefinitionProps
 ): string {
-  let templateInstance = JSON_TEMPLATE;
-
-  // env vars default is [], whereas secrets default is null. makes sense.
-  const envVarsValue = config.envVars ? JSON.stringify(config.envVars) : '[]';
-  const secretEnvVarsValue = config.secretEnvVars
-    ? JSON.stringify(config.secretEnvVars)
-    : 'null';
+  const containerDefinition = {
```
Comment (bassrock) [inline src/base/ApplicationECSContainerDefinition.ts:38]: 😄 yes. I'm glad you were able to make this work.

File: src/base/ApplicationECSContainerDefinition.ts [hunk 3]
```diff
@@ -92,81 +35,66 @@ export interface ApplicationECSContainerDefinitionProps {
...
-        command: config.healthCheck.command,
-        interval: config.healthCheck.interval,
-        retries: config.healthCheck.retries,
-        startPeriod: config.healthCheck.startPeriod,
-        timeout: config.healthCheck.timeout,
-      })
-    );
-  } else {
-    templateInstance = templateInstance.replace('[[HEALTH_CHECK]]', 'null');
+    containerDefinition['command'] = config.command;
```
Comment (bassrock) [inline src/base/ApplicationECSContainerDefinition.ts:96]: Do we leave this out if not defined? Interesting if so.. Can we comment on this?

**Review by bassrock — COMMENTED** [review]
this looks good to me, but I'm gonna delegate my approval to others.

File: src/base/ApplicationECSContainerDefinition.ts (hunk 3 above)
Comment (mmiermans) [reply src/base/ApplicationECSContainerDefinition.ts:96]: Good catch, I should've called this out for review. I did this to keep the behavior consistent, but I don't know why it was that way originally. I looked at the history and AWS docs:
- This comment from yourself is relevant, but doesn't reveal why the decision was made: https://github.com/Pocket/terraform-modules/pull/36#discussion_r515252198
- [The AWS docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-containerdefinitions.html#cfn-ecs-taskdefinition-containerdefinition-command) do not specify whether an empty list has different behavior than omitting this parameter.

Comment (mmiermans) [reply src/base/ApplicationECSContainerDefinition.ts:96]: So I found it odd, but I don't what comment to but there.

@jpetto, do you remember why `command` is removed from the json by default?

**Review by bassrock — APPROVED** [review]

**Review by kkyeboah — APPROVED** [review]
nice work 👍

**Pull request merged**